kube-core/src/labels.rs

```diff
@@ -0,0 +1,548 @@
+}
+
+/// A selector expression with existing operations
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub enum Expression {
+    /// Key exists and in set:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::In("foo".into(), ["bar".into(), "baz".into()].into()).to_string();
+    /// assert_eq!(exp, "foo in (bar,baz)")
+    /// ```
+    In(String, BTreeSet<String>),
+
+    /// Key does not exists or not in set:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotIn("foo".into(), ["bar".into(), "baz".into()].into()).to_string();
+    /// assert_eq!(exp, "foo notin (bar,baz)")
+    /// ```
+    NotIn(String, BTreeSet<String>),
+
+    /// Key exists and is equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Equal("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo=bar")
+    /// ```
+    Equal(String, String),
+
+    /// Key does not exists or is not equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotEqual("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo!=bar")
+    /// ```
+    NotEqual(String, String),
+
+    /// Key exists:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Exists("foo".into()).to_string();
+    /// assert_eq!(exp, "foo")
+    /// ```
+    Exists(String),
+
+    /// Key does not exist:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::DoesNotExist("foo".into()).to_string();
+    /// assert_eq!(exp, "!foo")
+    /// ```
+    DoesNotExist(String),
+
+    /// Invalid combination. Always evaluates to false.
+    Invalid,
+}
+
+/// Perform selection on a list of expressions
+#[derive(Clone, Debug, Eq, PartialEq, Default, Deserialize, Serialize)]
+pub struct Selector(Expressions);
+
+impl Selector {
+    /// Create a selector from a vector of expressions
+    fn from_expressions(exprs: Expressions) -> Self {
+        Self(exprs)
+    }
+
+    /// Create a selector from a map of key=value label matches
+    fn from_map(map: Map) -> Self {
+        Self(map.into_iter().map(|(k, v)| Expression::Equal(k, v)).collect())
+    }
+
+    /// Indicates whether this label selector matches all pods
+    pub fn selects_all(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Extend the list of expressions for the selector
+    ///
+    /// ```
+    /// use kube_core::Selector;
+    /// use kube_core::Expression;
+    /// use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+    ///
+    /// Selector::default().extend(Expression::Exists("foo".into()));
+    /// Selector::default().extend(Selector::from_iter(Expression::Exists("foo".into())));
+    /// let label_selector: Selector = LabelSelector::default().into();
+    /// Selector::default().extend(label_selector);
```

Comment:
maybe stick strictly to mutable extend here to fully show off the versatility of extend
```suggestion
    /// let mut selector = Selector::default();
    /// selector.extend(Expression::Equal("environment".into(), "production"));
    /// selector.extend([Expression::Exists("bar".into()), Expression::Exists("foo".into())].into_iter());
```
(assuming into_iter is the other direction of from_iter)

and in other ctors show of more natural instantiations

Reply:
Yeah, that’s better, should be fixed :)